src/sources/host_metrics/cgroups.rs

```diff
@@ -0,0 +1,271 @@
+use super::{filter_result_sync, FilterList, HostMetricsConfig};
+use crate::event::metric::Metric;
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+use shared::btreemap;
+use std::fs::{self, File};
+use std::io::{self, Read};
+use std::num::ParseIntError;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+#[derive(Clone, Debug, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+#[serde(default)]
+pub(super) struct CgroupsConfig {
+    #[derivative(Default(value = "3"))]
+    levels: usize,
+    base: Option<PathBuf>,
+    groups: FilterList,
+}
+
+impl HostMetricsConfig {
+    pub async fn cgroups_metrics(&self) -> Vec<Metric> {
+        let now = Utc::now();
+        CGroup::root(self.cgroups.base.as_deref())
+            .map(|root| self.recurse_cgroup(now, root, 1))
+            .unwrap_or_else(Vec::new)
+    }
+
+    fn recurse_cgroup(&self, now: DateTime<Utc>, cgroup: CGroup, level: usize) -> Vec<Metric> {
+        let mut result = Vec::new();
+
+        let tags = btreemap! {
+            "cgroup" => cgroup.name.to_string_lossy(),
+            "collector" => "cgroups",
+        };
+        if let Some(cpu) = filter_result_sync(
+            cgroup.load_cpu(),
+            "Failed to load/parse cgroups CPU statistics",
+        ) {
+            result.push(self.counter(
+                "cgroups_cpu_usage",
+                now,
+                cpu.usage_usec as f64 / 1_000_000.0,
+                tags.clone(),
+            ));
+            result.push(self.counter(
+                "cgroups_cpu_user",
+                now,
+                cpu.user_usec as f64 / 1_000_000.0,
+                tags.clone(),
+            ));
+            result.push(self.counter(
+                "cgroups_cpu_system",
+                now,
+                cpu.system_usec as f64 / 1_000_000.0,
+                tags.clone(),
+            ));
```

Comment:
Out of curiosity why not report the native numbers we receive here?

Reply:
Basically because "seconds" is a base unit (unscaled), and they have to be converted into `f64` anyways.

Reply:
Ah, the counter itself is in seconds unit?

Reply:
The underlying counter reported by the kernel is in microseconds. I decided to report it to consumers as seconds to match existing precedent (`cpu_seconds_total`).

Reply:
Gotcha. Absent other limitations my preference as a user would be to get the value without manipulation.